### PR TITLE
feat: Add default bytes and prefix

### LINKED ISSUE #148: Enable custom defaults for key prefix and bytes length
The current key creation form enables users to set a key prefix and bytes length (among others) with the default form values of "key" and "16" respectively. When creating many keys manually, this is annoying to keep changing these values to my custom overrides. It would reduce friction to enable saving a set of defaults either per API or per Workspace to help enforce consistency for those using the dashboard GUI.

### PR DESCRIPTION
**Pull request by harshsbhat**
## What does this PR do?

Allows to set default bytes and prefix for each API.

https://www.loom.com/share/ba99adb6585940f2be8834442a46210a?sid=ad746bca-d592-4533-b110-dda6b94bf0e9

Fixes #148

*If there is not an issue with this, please create one first. This is used for tracking purposes and also helps us understand why this PR exists*

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- Test A
- Test B

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [Contributing Guide](./CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand areas
- [x] Ran `pnpm build`
- [x] Ran `pnpm fmt`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Unkey Docs if changes were necessary

## Summary by CodeRabbit

- **New Features**
	- Introduced `DefaultBytes` and `DefaultPrefix` components for managing default settings of API keys.
	- Enhanced `CreateKey` component to accept new configuration options for bytes and prefix.
	- Updated settings page to include new components for API key management.
	- Added functionality for setting default byte sizes and prefixes through new API methods.
	- Implemented validation and submission handling for new settings components.

- **Bug Fixes**
	- Improved validation and error handling for API key settings.

- **Documentation**
	- Updated component descriptions and usage guidelines for new features.

### CODE REVIEW

**Comment by harshsbhat** [comment]
@perkinsjr this should be open for review now.

File: apps/dashboard/app/(app)/apis/[apiId]/settings/default-prefix.tsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    defaultValues: {
+      defaultPrefix: keyAuth.defaultPrefix ?? undefined,
+      keyAuthId: keyAuth.id,
+      workspaceId: keyAuth.workspaceId,
+    },
+  });
+
+  const setDefaultPrefix = trpc.api.setDefaultPrefix.useMutation({
+    onSuccess() {
+      toast.success("Default Prefix for this API is updated!");
```
Comment (chronark) [inline apps/dashboard/app/(app)/apis/[apiId]/settings/default-prefix.tsx:47]:
```suggestion
      toast.success("Default prefix for this API is updated!");
```

File: apps/dashboard/app/(app)/apis/[apiId]/settings/default-prefix.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      router.refresh();
+    },
+    onError(err) {
+      console.error(err);
+      toast.error(err.message);
+    },
+  });
+  async function onSubmit(values: z.infer<typeof formSchema>) {
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/dashboard/app/(app)/apis/[apiId]/settings/default-prefix.tsx:55]: yes please

File: apps/dashboard/app/(app)/apis/[apiId]/settings/default-prefix.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        <CardHeader>
+          <CardTitle>Default Prefix</CardTitle>
+          <CardDescription>
+            Set default Prefix for the keys under this API.
+          </CardDescription>
+        </CardHeader>
+        <CardContent>
+          <div className="flex flex-col space-y-2">
+            <input type="hidden" name="workspaceId" value={keyAuth.workspaceId} />
+            <input type="hidden" name="keyAuthId" value={keyAuth.id} />
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/dashboard/app/(app)/apis/[apiId]/settings/default-prefix.tsx:75]: no, this actually works

File: apps/dashboard/lib/trpc/routers/api/setDefaultPrefix.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+import { insertAuditLogs } from "@/lib/audit";
+import { db, eq, schema } from "@/lib/db";
+import { ingestAuditLogsTinybird } from "@/lib/tinybird";
+import { rateLimitedProcedure, ratelimit } from "@/lib/trpc/ratelimitProcedure";
+
+export const setDefaultApiPrefix = rateLimitedProcedure(ratelimit.update)
+  .input(
+    z.object({
+      defaultPrefix: z.string().max(8, "Prefix can be maximum of 8 charachters"),
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/dashboard/lib/trpc/routers/api/setDefaultPrefix.ts:12]: typo

File: apps/dashboard/lib/trpc/routers/api/setDefaultPrefix.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      workspaceId: z.string(),
+    }),
+  )
+  .mutation(async ({ ctx, input }) => {
+    const keyAuth = await db.query.keyAuth
+      .findFirst({
+        where: (table, { eq }) => eq(table.id, input.keyAuthId),
+        with: {
+          api: true,
+        },
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/dashboard/lib/trpc/routers/api/setDefaultPrefix.ts:23]: can you check if it's optional and remove?

File: apps/dashboard/lib/trpc/routers/api/setDefaultPrefix.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+          });
+        });
+      await insertAuditLogs(tx, {
+        workspaceId: keyAuth.workspaceId,
+        actor: {
+          type: "user",
+          id: ctx.user.id,
+        },
+        event: "api.update",
+        description: `Changed ${keyAuth.workspaceId} name from ${keyAuth.defaultPrefix} to ${input.defaultPrefix}`,
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/dashboard/lib/trpc/routers/api/setDefaultPrefix.ts:60]: correct, this is not updating a name

File: apps/dashboard/lib/trpc/routers/api/setDefaultPrefix.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      });
+    });
+    await ingestAuditLogsTinybird({
+      workspaceId: keyAuth.workspaceId,
+      actor: {
+        type: "user",
+        id: ctx.user.id,
+      },
+      event: "api.update",
+      description: `Changed ${keyAuth.id} name from ${keyAuth.defaultPrefix}} to ${input.defaultPrefix}`,
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/dashboard/lib/trpc/routers/api/setDefaultPrefix.ts:80]: same as above, we're not updating a name

**Comment by harshsbhat** [comment]
Should be good for review again @chronark

File: apps/dashboard/app/(app)/apis/[apiId]/settings/default-prefix.tsx (hunk 1 above)
Comment (chronark) [inline apps/dashboard/app/(app)/apis/[apiId]/settings/default-prefix.tsx:47]: this is still upper case

File: apps/dashboard/lib/trpc/routers/api/setDefaultBytes.ts [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      });
+    });
+    await ingestAuditLogsTinybird({
+      workspaceId: keyAuth.workspaceId,
+      actor: {
+        type: "user",
+        id: ctx.user.id,
+      },
+      event: "api.update",
+      description: `Changed ${keyAuth.id} default byte size for keys from ${keyAuth.defaultBytes}} to ${input.defaultBytes}`,
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/dashboard/lib/trpc/routers/api/setDefaultBytes.ts:81]: @harshsbhat did you see this?

**Review by chronark — CHANGES_REQUESTED** [review]

**Comment by harshsbhat** [comment]
I missed the extra bracket, there were two instances of capital P just fixed one and forgot about the other one. Should be good to review now tho

**Review by chronark — APPROVED** [review]

**Comment by harshsbhat** [comment]
Ah, I will need to resolve the conflicts here right?

**Comment by chronark** [comment]
yes please

**Comment by harshsbhat** [comment]
@chronark I have resolved the conflicts it was just two lines

**Review by chronark — APPROVED** [review]

File: apps/dashboard/lib/trpc/routers/api/setDefaultBytes.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (chronark) [reply apps/dashboard/lib/trpc/routers/api/setDefaultBytes.ts]: @harshsbhat please fix this

**Review by chronark — CHANGES_REQUESTED** [review]
still has an issue

**Comment by harshsbhat** [comment]
> still has an issue

@chronark its just one bracket. I am not sure how you are seeing two.

https://github.com/unkeyed/unkey/pull/2146/files#diff-7bff5ce27928c12e4e1e1213e7e9f6ca33aa58149bb34e28fb03ad227950c641R81

**Comment by chronark** [comment]
oh, I didn't see the "outdated" 😓 
CleanShot 2024-10-05 at 15 38 20@2x

**Review by chronark — APPROVED** [review]

**Review by chronark — APPROVED** [review]

**Pull request merged**